test/e2e/rollout_test.go

```diff
@@ -0,0 +1,119 @@
+// +build e2e
+
+/*
+Copyright 2021 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2e
+
+import (
+	"fmt"
+	"testing"
+
+	pkgtest "knative.dev/pkg/test"
+	"knative.dev/serving/pkg/apis/serving"
+	v1 "knative.dev/serving/pkg/apis/serving/v1"
+	testingv1 "knative.dev/serving/pkg/testing/v1"
+	"knative.dev/serving/test"
+	v1test "knative.dev/serving/test/v1"
+)
+
+// TestGradualRollout ensures the traffic is rolled out gradually.
+func TestGradualRollout(t *testing.T) {
+	t.Parallel()
+	clients := test.Setup(t)
+
+	names := test.ResourceNames{
+		Service: test.ObjectNameForTest(t),
+		Image:   test.PizzaPlanet1,
+	}
+
+	// Clean up on test failure or interrupt
+	test.EnsureTearDown(t, clients, &names)
+
+	// Setup Service
+	t.Log("Creating a new Service", names.Service)
+	robjs, err := v1test.CreateServiceReady(t, clients, &names,
+		testingv1.WithServiceAnnotation(serving.RolloutDurationKey, "60s"))
+	if err != nil {
+		t.Fatal("Create Ready Service:", err)
+	}
+
+	if _, err := v1test.PatchService(t, clients, robjs.Service, testingv1.WithServiceImage(pkgtest.ImagePath(test.Autoscale))); err != nil {
+		t.Fatalf("Patch update for Service %s with image %s failed: %v", names.Service, test.Autoscale, err)
+	}
+
+	sawRolloutStatus := false
+	sawTwoTrafficTargets := false
+
+	// This will cover all the status checks:
+	// - Status is in Rollout
+	// - Traffic is splitting
+	if err := v1test.WaitForServiceState(clients.ServingClient, names.Service,
+		func(s *v1.Service) (bool, error) {
+			// New revision not yet created.
+			if s.Status.LatestCreatedRevisionName == robjs.Revision.Name {
+				return false, nil
+			}
+
+			// Rollout not yet started.
+			if !sawRolloutStatus && s.Status.GetCondition(v1.ServiceConditionRoutesReady).GetReason() != "RolloutInProgress" {
+				return false, nil
+			}
+
+			// Rollout started!
+			if !sawRolloutStatus {
```

Comment:
I wonder if this test will be easier to follow if it were a sequence of `WaitForServiceState` 

ie. 
`WaitForServiceState(revisionCreated)`
`WaitForServiceState(sawTwoTrafficTargets)`
`WaitForServiceState(rolloutStatus)`
`WaitForServiceState(Ready)`

Reply:
I can separate first two. But the last two conditions are kind of intertwined.
Anyway, we can iterate on this. Given it's initial.
I have other ideas we can check later as well:
- duration
- that % change by the same amount each time
- etc

Reply:
But redid with several wait thingies.